java/ql/lib/semmle/code/java/frameworks/WebGoat.qll

```diff
@@ -0,0 +1,18 @@
+/**
+ * Provides models for the WebGoat Application.
+ */
+
+import java
+private import semmle.code.java.dataflow.ExternalFlow
+
+private class StringSummaryCsv extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        //`namespace; type; subtypes; name; signature; ext; input; output; kind`
+        "org.owasp.webgoat.assignments;AttackResult;false;AttackResult;;;Argument[1];Argument[-1];taint",
+        "org.owasp.webgoat.assignments;AttackResult;false;AttackResult;;;Argument[2];Argument[-1];taint",
+        "org.owasp.webgoat.assignments;AttackResult;false;AttackResult;;;Argument[3];Argument[-1];taint",
```

Comment:
I tried to model with `StringEscapeUtils.escapeJson()` and then `AttackResult` objects will indeed be tainted (partially). However, they will be tainted with an access path (to the tainted field assigned to in the constructor) and apparently this then does not work in the last step, where that object flows into a return statement. I am happy to learn about the way to make that work though.

Reply:
That's related to the modeling of the sink. Somehow Spring extracts the fields of `AttackResult`, so whichever rules leads us to model the sink, ought also to allow us to identify the relevant fields in QL. Once we have that, we can specify that these fields are implicitly read at the sink using `Configuration::allowImplicitRead(DataFlow::Node node, DataFlow::Content c)` (when overriding this predicate it is generally advisable to call super).
See e.g. the default implementation of `allowImplicitRead` on `TaintTracking::Configuration` - this specifies that array and collection contents can be implicitly read at sink nodes.

Reply:
Oh, I didn't know about this, thanks for pointing it out! Sounds very plausible indeed, so I'll wait what Tony has to say about the sinks.

Reply:
Agreed on adding `StringEscapeUtils.escapeJson` as a general taint step. The implicit read shouldn't be needed because a method returning an `AttackResult` isn't actually a XSS sink — see my comment below.